File: python/open_guardrail/guards/profanity_da.py
"""Detect Danish profanity."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"(?:^|\s)lort(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)pis(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)fanden(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)helvede(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)røv(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)luder(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)pikansjos(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)kraftedeme(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)skid(?:\s|$|[.,!?])", re.IGNORECASE),
    re.compile(r"(?:^|\s)idiot(?:\s|$|[.,!?])", re.IGNORECASE),
]


class _ProfanityDa:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "profanity-da"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []
        for pat in _PATTERNS:
            if pat.search(text):
                matched.append(pat.pattern)
        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="profanity-da",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message="Danish profanity detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def profanity_da(*, action: str = "block") -> _ProfanityDa:
    return _ProfanityDa(action=action)

File: packages/guards/src/profanity-da.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ProfanityDaOptions {
  action: 'block' | 'warn';
}

const PROFANITY_PATTERNS: RegExp[] = [
  /(?:^|\s)lort(?:\s|$|[.,!?])/i,
  /(?:^|\s)pis(?:\s|$|[.,!?])/i,
  /(?:^|\s)fanden(?:\s|$|[.,!?])/i,
  /(?:^|\s)helvede(?:\s|$|[.,!?])/i,
  /(?:^|\s)røv(?:\s|$|[.,!?])/i,
  /(?:^|\s)luder(?:\s|$|[.,!?])/i,
  /(?:^|\s)pikansjos(?:\s|$|[.,!?])/i,
  /(?:^|\s)kraftedeme(?:\s|$|[.,!?])/i,
  /(?:^|\s)skid(?:\s|$|[.,!?])/i,
  /(?:^|\s)idiot(?:\s|$|[.,!?])/i,
];

export function profanityDa(options: ProfanityDaOptions): Guard {
  return {
    name: 'profanity-da',
    version: '0.1.0',
    description: 'Detects Danish profanity',
    category: 'locale',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PROFANITY_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'profanity-da',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
